controllers/openstackplaybookgenerator_controller.go

```diff
@@ -114,6 +110,89 @@ func (r *OpenStackPlaybookGeneratorReconciler) Reconcile(ctx context.Context, re
 	templateParameters := make(map[string]interface{})
 	cmLabels := common.GetLabels(instance, openstackplaybookgenerator.AppLabel, map[string]string{})
 
+	// Check if fencing data is required as well (examine ControlPlane CR)
+	// FIXME: We assume there is only one ControlPlane CR for now (enforced by webhook), but this might need to change
+	controlPlaneList := &ospdirectorv1beta1.OpenStackControlPlaneList{}
+	controlPlaneListOpts := []client.ListOption{
+		client.InNamespace(instance.Namespace),
+		client.Limit(1000),
+	}
+	err = r.Client.List(context.TODO(), controlPlaneList, controlPlaneListOpts...)
+	if err != nil {
+		_ = r.setCurrentState(instance, ospdirectorv1beta1.PlaybookGeneratorError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	if len(controlPlaneList.Items) == 0 {
+		err := fmt.Errorf("no OpenStackControlPlanes found in namespace %s", instance.Namespace)
+		_ = r.setCurrentState(instance, ospdirectorv1beta1.PlaybookGeneratorError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// FIXME: See FIXME above
+	controlPlane := controlPlaneList.Items[0]
+
+	if controlPlane.Status.ProvisioningStatus.State != ospdirectorv1beta1.ControlPlaneProvisioned {
+		msg := fmt.Sprintf("Control plane %s VMs are not yet provisioned. Requeing...", controlPlane.Name)
+		r.Log.Info(msg)
+		err = r.setCurrentState(instance, ospdirectorv1beta1.PlaybookGeneratorWaiting, msg)
+		return ctrl.Result{RequeueAfter: time.Second * 10}, err
+	}
+
+	templateParameters["EnableFencing"] = false
+
+	if controlPlane.Spec.EnableFencing {
+		// TODO: This will likely need refactoring sooner or later
+		for roleName, roleParams := range controlPlane.Spec.VirtualMachineRoles {
+			if common.StringInSlice(roleParams.RoleName, controlplane.GetFencingRoles()) && roleParams.RoleCount == 3 {
+				// Get the associated VM instances
+				virtualMachineInstanceList, err := common.GetVirtualMachineInstances(r, instance.Namespace, map[string]string{
+					common.OwnerNameLabelSelector: roleName,
+				})
+
+				if err != nil {
+					_ = r.setCurrentState(instance, ospdirectorv1beta1.PlaybookGeneratorError, err.Error())
+					return ctrl.Result{}, err
+				}
+
+				fencingTemplateParameters, err := controlplane.CreateFencingConfigMapParams(virtualMachineInstanceList)
+
+				if err != nil {
+					_ = r.setCurrentState(instance, ospdirectorv1beta1.PlaybookGeneratorError, err.Error())
+					return ctrl.Result{}, err
+				}
+
+				templateParameters = common.MergeTemplateParameters(templateParameters, fencingTemplateParameters)
+				templateParameters["EnableFencing"] = true
+				// Only should be one Controller role, so break
+				break
```

Comment:
if you e.g. use ControllerOpenstack, which does not run rabbit and db, you'd also have VMs for Database and Messaging roles which also required fencing to be enabled. I know this is to add initial fencing config, just to keep in mind that if we have a role with 3 nodes+pacemaker service there might be also others.

Reply:
Ack, thanks for the feedback on this point.  Perhaps we remove the break and let the loop generate fencing device data for all VMs that belong to any role in `common/GetFencingRoles()`?

Reply:
yes, for the common.GetFencingRoles() it would be fine + checking for roles having `OS::TripleO::Services::Pacemaker` in any provided custom roles_data.yaml

Reply:
Can we do the checking of the provided custom roles in another follow-up PR?  I'd just like to get this one merged and then we can build on top of it.

Reply:
sure